### ISSUE TITLE
ExoPlayer does not recover from AudioTrack invalidation in passthrough [internal b/112299307] (AudioTrack write failed: -6)

### ISSUE THREAD

**Issue by Stonos** [google/ExoPlayer#7042]
### [REQUIRED] Issue description
I'm getting random `AudioSink$WriteException`s when playing a video with a 6 channel audio track (usually after pressing a button on the remote control).

This only seems to occur if the **HDMI Input Audio Format** setting on the TV is set to **Bitstream** instead of **PCM**.

I believe that this happens because Android is trying to play system sounds while in audio passthrough mode, because the following appeared in logcat after pressing a button on the remote control:
```AudioFlinger: openOutput() this 0xa56b7000, module 34 Device 8000, SamplingRate 48000, Format 0x000001, Channels 3, flags 0```

Here is the full stacktrace:
```
I AudioFlinger: openOutput() this 0xa56b7000, module 34 Device 8000, SamplingRate 48000, Format 0x000001, Channels 3, flags 0
D r_submix: adev_open_output_stream(address=-581846373:ap:1mixr:0)
D r_submix: submix_audio_device_create_pipe_l(addr=-581846373:ap:1mixr:0, idx=9)
D r_submix:   now using address -581846373:ap:1mixr:0 for route 9
I AudioFlinger: HAL output buffer size 1024 frames, normal sink buffer size 1024 frames
I AudioFlinger: AudioFlinger's thread 0xa4a83f00 tid=6091 ready to run
D AudioFlinger: ro.audio.silent will be ignored for threads on AUDIO_DEVICE_OUT_REMOTE_SUBMIX
I r_submix: out_standby()
D AudioFlinger: ro.audio.silent will be ignored for threads on AUDIO_DEVICE_OUT_REMOTE_SUBMIX
V AtvRemote.ABridgeX: AudioPolicy is registered in 16ms
E AudioTrack-JNI: Error -32 during AudioTrack native read
E ExoPlayerImplInternal: Renderer error: index=1, type=audio, format=Format(audio_ell=384000, null, null, audio/ac3, ac-3, 384000, el, [-1, -1, -1.0], [6, 48000]), rendererSupport=NO_UNSUPPORTED_DRM
E ExoPlayerImplInternal: com.google.android.exoplayer2.ExoPlaybackException: com.google.android.exoplayer2.audio.AudioSink$WriteException: AudioTrack write failed: -6
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.BaseRenderer.createRendererException(BaseRenderer.java:359)
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.processOutputBuffer(MediaCodecAudioRenderer.java:831)
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.drainOutputBuffer(MediaCodecRenderer.java:1585)
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:718)
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:599)
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:329)
E ExoPlayerImplInternal: 	at android.os.Handler.dispatchMessage(Handler.java:101)
E ExoPlayerImplInternal: 	at android.os.Looper.loop(Looper.java:164)
E ExoPlayerImplInternal: 	at android.os.HandlerThread.run(HandlerThread.java:65)
E ExoPlayerImplInternal: Caused by: com.google.android.exoplayer2.audio.AudioSink$WriteException: AudioTrack write failed: -6
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.audio.DefaultAudioSink.writeBuffer(DefaultAudioSink.java:772)
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.audio.DefaultAudioSink.handleBuffer(DefaultAudioSink.java:681)
E ExoPlayerImplInternal: 	at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.processOutputBuffer(MediaCodecAudioRenderer.java:824)
E ExoPlayerImplInternal: 	... 7 more
```

### [REQUIRED] Reproduction steps
1. Make sure the **HDMI Input Audio Format** setting (or equivalent) setting on your TV is set to **Bitstream**
2. Play a video that has a 6 channel audio track
3. Playback the video for about a minute
4. Press a few keys on the remote control. Playback should eventually stop with the error above.

### [REQUIRED] Link to test content
Provided via email.

### [REQUIRED] A full bug report captured from the device
Provided via email.

### [REQUIRED] Version of ExoPlayer being used
2.11.3

### [REQUIRED] Device(s) and version(s) of Android being used
Certified Android TV STB with Android 8.0 & Broadcom BCM7268 SoC.

**Comment by krocard** [comment]
This is an issue during a write of a passthrough AudioTrack. This could be for example because the AudioTrack was invalidated. For example: does using the Google Assistant from the remote trigger the bug (if your remote has a Google assistant button)?

As this is a low level platform error, this bug *can not be investigated further without a bug report* taken just after the write error. An additional bug report taken during a working playback would also be appreciated (start the same playback, start bug report generation, wait for ~3min without touching the remote and hopefully the bug triggering).

Note: -6 is -ENXIO No such device or address.

**Comment by Stonos** [comment]
Thanks for the reply.

I have already sent a bugreport to dev.exoplayer@gmail.com that was captured right after the issue occurred. Did you not receive it?

Tomorrow I will see if pressing the Google Assistant key reproduces the issue. I will also grab a bugreport after 3 minutes of normal playback.

**Comment by krocard** [comment]
Just found your email! Will take a look.

**Comment by krocard** [comment]
As suspected, there was an audio routing event just before the write failure:
```
02-28 17:55:10.493  1041  3463  3589 D r_submix: adev_open_output_stream(address=-581846373:ap:1mixr:0)
02-28 17:55:10.493  1041  3463  3589 D r_submix: submix_audio_device_create_pipe_l(addr=-581846373:ap:1mixr:0, idx=9)
02-28 17:55:10.506 u0_a1  5338  5363 V AtvRemote.ABridgeX: AudioPolicy is registered in 16ms
```
This is similar to the internal issue b/119801918.

1. A system application registers/unregisters a dynamic AudioPolicy.
2. Some AudioTrack are invalidated, including the Exoplayer one.
3. Exoplayer receives a write failure.

This is what we can do about it:

1. Nothing: AudioPolicy are used for many reason, in general to control the audio routing by system Application, but also by the AudioPlaybackCapture API.
2. In P and Q, the Android Audio framework team has improved the invalidation logic, so this bug will trigger less frequently on recent TVs. This bug report was taken on O, which invalidates all track on AP register/unregister.
3. Most importantly, ExoPlayer should better handle AudioTrack invalidation.

ExoPlayer original assumption that its audio track will always stay valid is no longer true. Track invalidation used to be rare, but are more and more frequent with system feature as such as TV remote audio, Auto casting, Audio Playback Capture...

AudioTrack being a low level API, does not hide it's invalidations from it's client even if sometime it would be possible. (In that it is similar to AAudio).

To conclude, this bug can not be easily solved, ExoPlayer needs to reopen and restart an AudioTrack on such errors. Unfortunately that will still not be invisible for the user. This work is already planned for Offload but needs to be expanded for all AudioTrack.

On the bright side, those errors should be much more rare on more recent Android version.

**Comment by krocard** [comment]
> I will also grab a bugreport after 3 minutes of normal playback.

No need, thanks. The first one was enough. I would still be interested if you could test the Assistant as a nice way to reproduce if needed.

**Comment by Stonos** [comment]
Unfortunately it's very unlikely that the STB I'm working with will ever receive an update to Android 9/10. Is there anything that could be done in order to reduce the chance of triggering this issue in the first place?

**Comment by andrewlewis** [comment]
For the record: [Internal: b/112299307] is the corresponding internal bug for improving handling of the audio track being invalidated.

**Comment by Stonos** [comment]
Update: I tried pressing the assistant button while playing the same content and the bug did not trigger (i.e. the video continued playing normally). However, I couldn't hear the assistant talking. Once I switched to the stereo channel the assistant was audible.

**Comment by krocard** [comment]
This bug will track the implementation of AudioTrack invalidation recovery.

**Comment by krocard** [comment]
Audio track retry has been implemented in ExoPlayer ~v2.12.0~ and will be release in v2.13*. ExoPlayer will now attempt to recreate a track after it fails to be created for 100ms.
Please reopen if your issue is not fixed.

EDIT: The retry did not make it in 2.12.0, see https://github.com/google/ExoPlayer/issues/7042#issuecomment-707062931

**Comment by krocard** [comment]
The commit that implements the retry is: 7dfdde9.

Nevertheless, the commit is not part of v2.12 as I have previously written. It was merged after the branch cut.
Thus it will be part of the next release 2.13.0 that should be published at the beginning of next year.

Sorry for the confusion. You can test the fix by either checkout the commit or our development branch `dev-v2`.
